exchanges/gateio/gateio_wrapper.go

```diff
@@ -208,90 +221,424 @@ func (g *Gateio) Run() {
+func (g *Gateio) FetchTradablePairs(ctx context.Context, a asset.Item) (currency.Pairs, error) {
+	if !g.SupportsAsset(a) {
+		return nil, fmt.Errorf("%s does not support %s", g.Name, a)
 	}
-	for p := range pairs {
-		for k := range result {
-			if !strings.EqualFold(k, pairs[p].String()) {
+	switch a {
+	case asset.Spot:
+		tradables, err := g.ListAllCurrencyPairs(ctx)
+		if err != nil {
+			return nil, err
+		}
+		pairs := make([]currency.Pair, 0, len(tradables))
+		for x := range tradables {
+			p := strings.ToUpper(tradables[x].Base + currency.UnderscoreDelimiter + tradables[x].Quote)
+			if !g.IsValidPairString(p) {
 				continue
 			}
-
-			err = ticker.ProcessTicker(&ticker.Price{
-				Last:         result[k].Last,
-				High:         result[k].High,
-				Low:          result[k].Low,
-				Volume:       result[k].BaseVolume,
-				QuoteVolume:  result[k].QuoteVolume,
-				Open:         result[k].Open,
-				Close:        result[k].Close,
-				Pair:         pairs[p],
-				ExchangeName: g.Name,
-				AssetType:    a})
+			cp, err := currency.NewPairFromString(p)
 			if err != nil {
-				return err
+				return nil, err
+			}
+			pairs = append(pairs, cp)
+		}
+		return pairs, nil
+	case asset.Margin, asset.CrossMargin:
+		tradables, err := g.GetMarginSupportedCurrencyPairs(ctx)
+		if err != nil {
+			return nil, err
+		}
+		pairs := []currency.Pair{}
+		for x := range tradables {
+			p := strings.ToUpper(tradables[x].Base + currency.UnderscoreDelimiter + tradables[x].Quote)
+			if !g.IsValidPairString(p) {
+				continue
+			}
+			cp, err := currency.NewPairFromString(p)
+			if err != nil {
+				return nil, err
+			}
+			pairs = append(pairs, cp)
+		}
+		return pairs, nil
+	case asset.Futures:
+		btcContracts, err := g.GetAllFutureContracts(ctx, settleBTC)
+		if err != nil {
+			return nil, err
+		}
+		usdtContracts, err := g.GetAllFutureContracts(ctx, settleUSDT)
+		if err != nil {
+			return nil, err
+		}
+		btcContracts = append(btcContracts, usdtContracts...)
+		pairs := []currency.Pair{}
+		for x := range btcContracts {
+			p := strings.ToUpper(btcContracts[x].Name)
+			if !g.IsValidPairString(p) {
+				continue
+			}
+			cp, err := currency.NewPairFromString(p)
+			if err != nil {
+				return nil, err
+			}
+			pairs = append(pairs, cp)
+		}
+		return pairs, nil
+	case asset.DeliveryFutures:
+		btcContracts, err := g.GetAllDeliveryContracts(ctx, settleBTC)
+		if err != nil && !strings.Contains(err.Error(), "404 Not Found") {
+			return nil, err
+		}
+		usdContracts, err := g.GetAllDeliveryContracts(ctx, settleUSD)
+		if err != nil && !strings.Contains(err.Error(), "404 Not Found") {
+			return nil, err
+		}
+		usdtContracts, err := g.GetAllDeliveryContracts(ctx, settleUSDT)
+		if err != nil && !strings.Contains(err.Error(), "404 Not Found") {
+			return nil, err
+		}
+		btcContracts = append(btcContracts, usdtContracts...)
+		btcContracts = append(btcContracts, usdContracts...)
+		pairs := []currency.Pair{}
```

Comment:
```suggestion
		pairs := make([]currency.Pair, 0, btcContracts)
```
Main reason is that when declaring an empty slice, its nicer to just go `var pairs []currency.Pair`, but then since we know here the rough size of the slice, we can actually declare its length like above